install/scripts/craycli_init.py

```diff
@@ -0,0 +1,753 @@
+                sys.exit(1)
+
+        # pull the values from the secret if present
+        if sec != None:
+            self._tu_username = base64.b64decode(sec.data['user']).decode('ascii')
+            self._tu_password = base64.b64decode(sec.data['password']).decode('ascii')
+        return
+
+# Class to wrap keycloak authentication with calls to the keycloak rest api
+class KeycloakSetup(object):
+    MASTER_REALM_NAME = 'master'
+    SHASTA_REALM_NAME = 'shasta'
+
+    def __init__(
+            self,
+            keycloak_base,
+            kc_master_admin_client_id,
+            kc_master_admin_username,
+            kc_master_admin_password):
+        self.keycloak_base = keycloak_base
+        self.kc_master_admin_client_id = kc_master_admin_client_id
+        self.kc_master_admin_username = kc_master_admin_username
+        self.kc_master_admin_password = kc_master_admin_password
+
+        self._kc_master_admin_client_cache = None
+
+    # creates/gets a client object with the correct auth tokens present
+    @property
+    def _kc_master_admin_client(self):
+        if self._kc_master_admin_client_cache:
+            return self._kc_master_admin_client_cache
+
+        kc_master_token_endpoint = (
+            '{}/realms/{}/protocol/openid-connect/token'.format(
+                self.keycloak_base, self.MASTER_REALM_NAME))
+
+        kc_master_client = oauthlib.oauth2.LegacyApplicationClient(
+            client_id=self.kc_master_admin_client_id)
+
+        client = requests_oauthlib.OAuth2Session(
+            client=kc_master_client, auto_refresh_url=kc_master_token_endpoint,
+            auto_refresh_kwargs={
+                'client_id': self.kc_master_admin_client_id,
+            },
+            token_updater=lambda t: LOGGER.info("Refreshed Keycloak master admin token"))
+        LOGGER.debug("Fetching initial KC master admin token.")
+        client.fetch_token(
+            token_url=kc_master_token_endpoint,
+            client_id=self.kc_master_admin_client_id,
+            username=self.kc_master_admin_username,
+            password=self.kc_master_admin_password)
+
+        self._kc_master_admin_client_cache = client
+        return self._kc_master_admin_client_cache
+
+    # Create the user in keycloak with the given password
+    def create_user(self, username, password, group):
+        # check for valid input
+        if username == "" or password == "":
+            LOGGER.error(f"May not create user with empty name or password")
+            return
+
+        # construct the url and request body to create the user
+        createUserUrl = f"{self.keycloak_base}/admin/realms/{self.SHASTA_REALM_NAME}/users"
+        req_body = {
+            'username': username,
+            'enabled': True,
+            'credentials': [
+                {
+                    'type': 'password',
+                    'value': password,
+                },
+            ],
+            'groups': [group,]
```

Comment:
Should there be a comma in there?

Reply:
Yes, but I would be a little hard-pressed to explain why...